mgr/dashboard: unify button/URL actions naming

epuertat opened the pull request:
The screenshots below show an example of the changes performed. They've been applied to the top 8 components (Pools, RBD images, iSCSI, mirroring, RGW users, RGW buckets, Users and Roles).

Before:
![before](https://user-images.githubusercontent.com/37327689/55000423-c2b6d800-4fd2-11e9-98b8-543986c69b55.png)

After:
![after](https://user-images.githubusercontent.com/37327689/55000485-ea0da500-4fd2-11e9-8ba3-f412ceb28c5b.png)


Detailed changes:
- Mappings (actually an Enum) created for actions (buttons and other UI elements) and URLs: `ActionLabels` and `URLVerbs`.
  - Probably a generic store for translatable strings would be better.
  - An alternative would be to fix/improve the current i18n-polyfill, which only works with literal strings (not even with 'const enums' which become literals after Typescript transpiling).
  - Additionally having a predefined file with some strings to translate (actions, verbs, etc) could improve on the 1st of the 2-stage i18n process (as extraction tool has a lot of limitations).
- A corresponding `ActionLabelsI18n` service with translated labels (it's a service as I haven't found the way to either translate no-const strings (ngx-translate/AST parser failure) or get a static translator).
  - This services could/should be extended to cover all strings that are defined in static/globally scoped objects before any I18n provider has been initialized.
- Breadcrumbs are not translated (neither were they before this change). This part remains untackled: using 'proxy' static objects and performing live translation could deal with the issue.
- New `URLBuilder` service created (following a established pattern in the Java/.NET world) . This should avoid the need of messing with literal URLs and string composition/parsing, and while the front-end is not meant to be consumed by anyone, Angular does not provide any other way for the app to navigate between components, so the URLs are a de-facto interface contract. Unlike this approach is not flawless, it's easier to enforce, while issues coming from free-from strings are really hard to catch.
  - This could be further improved by using a router registry/dynamic routing. Most of the routes are trivial.
- This PR initially included some code to remove leading/trailing whitespace/newlines from i18n elements in XLIFF file, but it turned out it didn't work correctly as long as the extractor tool will still calculate different IDs for equal strings with varying whitespacing.
- As a side effect of these changes, routing module has been refactored and some routes moved to their specific modules (pool, rbd, rgw), via `loadChildren` and `routes.forChild()` magic. Now the above mentioned components are lazy-loaded/pre-loaded (it means right after the main code is loaded). This should also decrease the loading time (though probably this is not biggest time eater here).
  - As now modules can be loaded multiple times, not only from App module by means of lazy loading, but also from other ones (as PoolModule loads BlockModule to get QoS widgets in Pool windows), now lazy loaded modules include 2 `NgModules` (one with `imports: RouterModule.forChild(routes)`, meant for lazy-loading, and another without routes).
- Some parts might not be (fully) translated (NFS, iSCSI, mirroring), as there's been ongoing work on them and it's hard to keep up with the new code.
- These changes will be a waste of time if the new code does not take benefit from/adheres to it, so I'm still figuring out how to spread this (nothing really fancy to demo). Maybe adding some checks/harnessing to enforce the new naming convention (ideas greatly welcome here).

Fixes: http://tracker.ceph.com/issues/37337
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-details/pool-details.component.spec.ts:
```diff
@@ -14,7 +14,7 @@ describe('PoolDetailsComponent', () => {
   configureTestBed({
     imports: [TabsModule.forRoot(), AppModule],
-    decalarations: [PoolDetailsComponent],
+    declarations: [PoolDetailsComponent],
```
🙊

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -295,12 +295,12 @@ <h3 class="panel-title">
                             type="button"
                             (submitAction)="submit()">
-            <span i18n>{mode, select, editing {Update} cloning {Clone} copying {Copy} other {Create}}</span> RBD
+            <span i18n>{mode, select, editing {{{ActionLabels.EDIT}}} cloning {{{ActionLabels.CLONE}}} copying {{{ActionLabels.COPY}}} other {{{ActionLabels.CREATE}}}}</span> image
```
I think that change is not properly parsed by ng-xi18n (instead of the `enum` literal expansion it's hitting the symbol itself). An alternative could be to treat that string in the TS side, rather than in the template.

Lots of errors are printed like this:
```ts
# ng xi18n --i18n-format xlf --i18n-locale en-US --output-path locale --progress=false && ngx-extractor -i 'src/**/*.ts' -f xlf -o src/locale/messages.xlf -l en-US

SKIP: Unknown argument type: 'PropertyAccessExpression' NodeObject {
  pos: 4159,
  end: 4178,
  flags: 0,
  transformFlags: undefined,
  parent: undefined,
  kind: 189,
  expression:
   IdentifierObject {
     pos: 4159,
     end: 4171,
     flags: 0,
     parent: undefined,
     escapedText: 'ActionLabels' },
  name:
   IdentifierObject {
     pos: 4172,
     end: 4178,
     flags: 0,
     parent: undefined,
     escapedText: 'CREATE' },
  _children:
   [ IdentifierObject {
       pos: 4159,
       end: 4171,
       flags: 0,
       parent: undefined,
       escapedText: 'ActionLabels' },
     TokenObject { pos: 4171, end: 4172, flags: 0, parent: [Circular], kind: 24 },
     IdentifierObject {
       pos: 4172,
       end: 4178,
       flags: 0,
       parent: undefined,
       escapedText: 'CREATE' } ] }
```

epuertat replied:
It seems it has nothing to do with this but with a limitation in `ngx-extractor` to deal with some AST nodes, specifically 
`PropertyAccessExpression`:

https://github.com/ngx-translate/i18n-polyfill/blob/1db63d046d5a356570a259874e2e08ca9a0e4b2e/lib/extractor/src/abstract-ast-parser.ts#L20-L42

epuertat commented:
jenkins test make check

votdev commented:
Could you please rebase this PR, otherwise there is a merge odyssey when you pull this PR into master.

epuertat commented:
> Could you please rebase this PR, otherwise there is a merge odyssey when you pull this PR into master.

Sure. Tried late yesterday but the amount of changes was... awful.

votdev commented:
> > Could you please rebase this PR, otherwise there is a merge odyssey when you pull this PR into master.
> 
> Sure. Tried late yesterday but the amount of changes was... awful.

Still have merge conflicts.

```
$ git fetch upstream
$ git rebase upstream/master
$ git checkout -b testing
$ git pull upstream pull/26572/head
...
Applying: qa/workunits/rbd: add parent migration case
Applying: objclass: fix cls_get_snapset_seq for cache tier case
Applying: common: define BOOST_COROUTINES_NO_DEPRECATION_WARNING if not yet
Applying: mgr: Ignore __pycache__ and wheelhouse dirs
Applying: CLI: ability to change file ownership
Applying: mgr/dashboard: Add support for managing RBD QoS
Using index info to reconstruct a base tree...
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.spec.ts
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
M	src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
.git/rebase-apply/patch:1245: new blank line at EOF.
+
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.spec.ts
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
error: Failed to merge in the changes.
Patch failed at 0183 mgr/dashboard: Add support for managing RBD QoS
Use 'git am --show-current-patch' to see the failed patch
```

epuertat commented:
> > > Could you please rebase this PR, otherwise there is a merge odyssey when you pull this PR into master.
> > 
> > 
> > Sure. Tried late yesterday but the amount of changes was... awful.
> 
> Still have merge conflicts.
> 
> ```
> $ git fetch upstream
> $ git rebase upstream/master
> $ git checkout -b testing
> $ git pull upstream pull/26572/head
> ...
> Applying: qa/workunits/rbd: add parent migration case
> Applying: objclass: fix cls_get_snapset_seq for cache tier case
> Applying: common: define BOOST_COROUTINES_NO_DEPRECATION_WARNING if not yet
> Applying: mgr: Ignore __pycache__ and wheelhouse dirs
> Applying: CLI: ability to change file ownership
> Applying: mgr/dashboard: Add support for managing RBD QoS
> Using index info to reconstruct a base tree...
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.spec.ts
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
> M	src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
> M	src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
> .git/rebase-apply/patch:1245: new blank line at EOF.
> +
> warning: 1 line adds whitespace errors.
> Falling back to patching base and 3-way merge...
> Auto-merging src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool.module.ts
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.spec.ts
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html
> Auto-merging src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
> CONFLICT (content): Merge conflict in src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts
> error: Failed to merge in the changes.
> Patch failed at 0183 mgr/dashboard: Add support for managing RBD QoS
> Use 'git am --show-current-patch' to see the failed patch
> ```

Yeah, sorry, I just posted the rebased PR. That one you tried was unrelated to that.

epuertat commented:
jenkins test make check

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/erasure-code-profile-form/erasure-code-profile-form.component.html:
```diff
@@ -323,10 +323,10 @@
             type="button"
             (click)="bsModalRef.hide()"
-            i18n>Close</button>
+            i18n>{{ ActionLabels.CANCEL }}</button>
```
```suggestion
            >{{ ActionLabels.CANCEL }}</button>
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/erasure-code-profile-form/erasure-code-profile-form.component.ts:
```diff
@@ -28,6 +29,7 @@ export class ErasureCodeProfileFormComponent implements OnInit {
   requiredControls: string[] = [];
   devices: string[] = [];
   tooltips = this.ecpService.formTooltips;
+  readonly ActionLabels = ActionLabels;
```
Should use `ActionLabelsI18n`.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -452,12 +452,12 @@ <h3 class="panel-title">
-            <span i18n>{editing, select, 1 {Edit} other {Create}} pool</span>
+            <span i18n>{editing, select, 1 {{{ActionLabels.EDIT}}} other {{{ActionLabels.CREATE}}}} Pool</span>
           </cd-submit-button>
           <button i18n
```
```suggestion
           <button 
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts:
```diff
@@ -62,6 +63,7 @@ export class PoolFormComponent implements OnInit {
     sourceType: RbdConfigurationSourceField;
   }>();
   currentConfigurationValues: { [configKey: string]: any } = {};
+  readonly ActionLabels = ActionLabels;
```
You should use `ActionLabelsI18n`.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:
```diff
@@ -99,12 +99,12 @@
-            <ng-container i18n>{editing, select, 1 {Update} other {Add}}</ng-container>
+            <ng-container i18n>{editing, select, 1 {{{ActionLabels.EDIT}}} other {{{ActionLabels.CREATE}}}} Bucket</ng-container>
           </cd-submit-button>
           <button i18n
```
```suggestion
           <button 
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:
```diff
@@ -23,6 +24,7 @@ export class RgwBucketFormComponent implements OnInit {
   error = false;
   loading = false;
   owners = null;
+  readonly ActionLabels = ActionLabels;
```
You should use `ActionLabelsI18n`

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.html:
```diff
@@ -590,12 +590,12 @@ <h3 class="panel-title">{editing, select, 1 {Edit} other {Add}} user</h3>
                   type="button"
                   routerLink="/rgw/user"
-                  i18n>Back</button>
+                  i18n>{{ActionLabels.CANCEL}}</button>
```
```suggestion
                  >{{ActionLabels.CANCEL}}</button>
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts:
```diff
@@ -40,6 +41,8 @@ export class RgwUserFormComponent implements OnInit {
   swiftKeys: RgwUserSwiftKey[] = [];
   capabilities: RgwUserCapability[] = [];
 
+  readonly ActionLabels = ActionLabels;
```
You should use `ActionLabelsI18n`

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html:
```diff
@@ -78,11 +78,11 @@
                             (submitAction)="submit()"
-                            i18n>{mode, select, editing {Update} other {Create}} Role</cd-submit-button>
+                            i18n>{mode, select, editing {{{ActionLabels.EDIT}}} other {{{ActionLabels.CREATE}}}} Role</cd-submit-button>
           <button i18n
```
```suggestion
           <button 
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts:
```diff
@@ -42,6 +43,8 @@ export class RoleFormComponent implements OnInit {
   roleFormMode = RoleFormMode;
   mode: RoleFormMode;
 
+  readonly ActionLabels = ActionLabels;
```
You should use `ActionLabelsI18n`.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.html:
```diff
@@ -144,11 +144,11 @@
                             (submitAction)="submit()"
-                            i18n>{mode, select, editing {Update} other {Create}} User</cd-submit-button>
+                            i18n>{mode, select, editing {{{ActionLabels.EDIT}}} other {{{ActionLabels.CREATE}}}} User</cd-submit-button>
           <button i18n
```
```suggestion
           <button 
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts:
```diff
@@ -39,6 +40,8 @@ export class UserFormComponent implements OnInit {
   allRoles: Array<UserFormRoleModel>;
   messages = new SelectMessages({ empty: 'There are no roles.' }, this.i18n);
 
+  readonly ActionLabels = ActionLabels;
```
You should use `ActionLabelsI18n`.

epuertat replied:
Thanks a lot, @tspmelo, for your thorough review! I'm currently fixing all these issues, so I'll let you know when they are pushed.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -7,9 +7,8 @@
-            i18n>{mode, select, editing {Edit} cloning {Clone} copying {Copy} other {Add}}</span>
-          RBD
+          <span i18n>{mode, select, editing {{{ActionLabels.EDIT}}} cloning {{{ActionLabels.CLONE}}} copying {{{ActionLabels.COPY}}} other {{{ActionLabels.CREATE}}}}</span>
+          Image
```
Thanks, @tspmelo! Yup, that's exactly what I was doing right now (@alfonsomthd warned me about this):
```html
<span i18n="@@formTitle">{{action | titlecase}} {{resource | titlecase}}</span>
```

And feeding translated `action` and `resource`.

tspmelo replied:
That might look strange in transifex, a translation with just 2 untranslatable blocks, but it will be useful for languages that change the order of the sentence.

epuertat replied:
I was wondering exactly the same: by leaving i18n on the span block, the whole string can still be accessible for a translator. For making this clear, I'll add a i18n description to aid on the translation.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -7,9 +7,8 @@
-          <span
-            i18n>{mode, select, editing {Edit} cloning {Clone} copying {Copy} other {Add}}</span>
-          RBD
+          <span i18n>{mode, select, editing {{{ActionLabels.EDIT}}} cloning {{{ActionLabels.CLONE}}} copying {{{ActionLabels.COPY}}} other {{{ActionLabels.CREATE}}}}</span>
```
I tried Angular `i18nSelect` pipe but it improves zero improvement compared to this. So I'll handle the `select` from the TS side, and provide a `<span i18n>{{action}} {{resource}}</span>`.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -310,14 +309,14 @@ <h3 class="page-header" i18n>Striping</h3>
                   class="btn btn-sm btn-default"
                   routerLink="/block/rbd"
-                  i18n>Back</button>
+                  i18n>{{ActionLabels.CANCEL}}</button>
```
Yup, I suffered this one too. I'm using now `ActionLabelsI18n`, but the good thing is that removing the `i18n` is not required for this to work (just checked).

epuertat commented:
jenkins test docs

tspmelo commented on src/pybind/mgr/dashboard/frontend/package.json:
```diff
@@ -2,12 +2,13 @@
     "build": "npm run env_build && ng build",
     "env_build": "cp src/environments/environment.tpl.ts src/environments/environment.prod.ts && cp src/environments/environment.tpl.ts src/environments/environment.ts  && node ./environment.build.js",
-    "i18n": "ng xi18n --i18n-format xlf --i18n-locale en-US --output-path locale --progress=false && ngx-extractor -i 'src/**/*.ts' -f xlf -o src/locale/messages.xlf -l en-US",
+    "i18n": "ng xi18n --i18n-format xlf --i18n-locale en-US --output-path locale --progress=false && ngx-extractor -i 'src/**/*.ts' -f xlf -o $npm_package_locale -l en-US && perl -0777 -i -pe 's|(<source>)[\\s]+(.*?)[\\s]+(</source>)|$1$2$3|g' $npm_package_locale",
```
After reading the new description I remembered why this is not a real solution to the problem.
If you write the following html:
```
<ng-container i18n>ASD</ng-container>
<ng-container i18n> ASD</ng-container>
<ng-container i18n>
  ASD
</ng-container>
```
You will still get 3 `trans-units`, each with a different id, but with the same string.

The solution to this problem would be to either run a script to fix the source code or fix the i18n extractor itself.

epuertat replied:
Sh*t! You're totally right! And the 'misbehaviour' is not only on the polyfill part, the standard Angular i18n creates new translation units for every whitespace change (even in interpolated parts, which is nonsensical). So... I'm afraid the only possible solution is: either using linters to enforce adherence to coding guidelines, or (my preferred) move the HTML formatting script to the html code. I'll look at this.

As a workaround in the meantime, by specifying named ids `< ... i18n="..@@named_id" ...>` we might temporarily avoid this issue.

epuertat replied:
So, as agreed, I'll remove the i18n perl regex script and fix the newlines within i18n elements.

To follow the current status of Angular i18n implementation, let's keep an eye on https://github.com/angular/angular/issues/16477

epuertat replied:
Btw, [these are the lines](https://github.com/angular/angular/blob/f8c7c3c09c38089558aa95a1f9c5a297fed852a6/packages/compiler/src/i18n/digest.ts#L13-L15)  that generate the trans-unit ID:
```ts
export function digest(message: i18n.Message): string {
  return message.id || sha1(serializeNodes(message.nodes).join('') + `[${message.meaning}]`);
}
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:
```diff
@@ -12,8 +12,8 @@
       <div class="panel-heading">
-        <h3 class="panel-title">
-          <span i18n>{editing, select, 1 {Edit} other {Add}} pool</span>
+        <h3 class="panel-title" i18n="form title|Example: Create Pool@@formTitle">
```
I would suggest you remove the `@@formTitle` or use the same meaning for both `formTitle` and `formActionButton`. 
Currently we get 2 different `trans-units`, just because the meaning is different.

epuertat replied:
Not so sure about this. Why? I'm using `{{ action | titlecase }} {{ resource | titlecase }}` for both a descriptive element (Form title: "Create Pool") and for imperative one (Button label: "Create Pool"), and I chose the `titlecase` capitalization because I was following this [recommendation](https://www.patternfly.org/styles/terminology-and-wording/#capitalization), which only takes English grammar into consideration.

Other languages might follow different practices (e.g.: nouns in German are always in uppercase, while in Spanish headline capitalization is incorrect, and I cannot imagine what non-indoeuropean languages would do about this). I prefer to keep room for a translator to deal with each case separately.

tspmelo replied:
Transifex doesn't allow you to change the interpolations.
The translator will only see `<1> <2>` and can only change the order of the 2 words, they will not be able to change the capitalization.
p.e.: https://www.transifex.com/ceph/ceph-dashboard/translate/#pt_PT/messagesxlf/164387632

epuertat replied:
Oops, true! I also checked in [the ICU](http://userguide.icu-project.org/transforms/casemappings) for the case the had some way to deal with that, but I don't see anything specific to Titlecase/Headings.

However, I still prefer to keep 2 trans-units for this (even if they are _right now_ exactly the same), as they semantically 2 different things. With this change we have already reduced 2 x 2 x 8 (32) translation units to just 2.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/iscsi-target-form/iscsi-target-form.component.html:
```diff
@@ -559,12 +560,13 @@
                   class="btn btn-sm btn-default"
                   routerLink="/block/iscsi/targets"
-                  i18n>Back</button>
+                  i18n>{{ actionLabels.CANCEL }}</button>
```
Thanks for noticing this. This is fixed in my last update.

epuertat commented:
@tspmelo do you know if we have `preserveWhitespace` enabled? I assumed Angular >=6 had it disabled, so whitespace should be removed (something that might help sort out this issue), but just checked and it remains (at least within text elements).

tspmelo commented:
@epuertat I think we didn't change the default configuration. But I'm guessing that only applies for the final html render.

epuertat commented:
> @epuertat I think we didn't change the default configuration. But I'm guessing that only applies for the final html render.

I checked both dev and prod builds, and at least prod builds seem to remove all indentation whitespaces, but duplicate whitespace within text-related tags is kept (dev builds seem to bypass all HTML whitespacing).

epuertat commented:
jenkins test make check

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/iscsi-target-form/iscsi-target-form.component.ts:
```diff
@@ -74,6 +75,8 @@ export class IscsiTargetFormComponent implements OnInit {
   USER_REGEX = /[\w\.:@_-]{8,64}/;
   PASSWORD_REGEX = /[\w@\-_\/]{12,16}/;
+  action: string;
+  resource = 'target';
```
This needs to be translated.
Either translate it in the TS or remove the variable and translate in the HTML.

Same applies to all other `resource` vars.

epuertat replied:
I initially assumed that probably most 'resources' would go untranslated, as they were very so specific that would remain in 'English' (Pool, bucket, RBD image), but others are easier (e.g.: user). So I'll do that.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/iscsi-target-form/iscsi-target-form.component.html:
```diff
@@ -8,7 +8,8 @@
         <h3 class="panel-title"
-            i18n>{isEdit, select, 1 {Edit} other {Add}} target</h3>
+            i18n="form title|Example: Create Pool@@formTitle"
+            >{{ action | titlecase }} {{ resource | titlecase }}</h3>
```
We decided not to use the `>` in a separate line from the attrs.

epuertat replied:
I've used now the attribute alignment, but given the weirdness we're dealing with (i18n and whitespacing), I think the trailing `>` is a minor issue :stuck_out_tongue:

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts:
```diff
@@ -85,6 +86,8 @@ export class RbdFormComponent implements OnInit {
     '32 MiB'
   ];
+  action: string;
+  resource = 'rbd';
```
I think this should allways be all caps: `RBD`

epuertat replied:
Done.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.html:
```diff
@@ -589,13 +591,13 @@ <h3 class="panel-title">{editing, select, 1 {Edit} other {Add}} user</h3>
-                            [form]="userForm">
-            <ng-container i18n>{editing, select, 1 {Update} other {Add}}</ng-container>
-          </cd-submit-button>
+                            [form]="userForm"i18n="form action button|Example: Create Pool@@formActionButton"
```
```suggestion
                            [form]="userForm"
                            i18n="form action button|Example: Create Pool@@formActionButton"
```

epuertat replied:
Great catch! This passed all the linters and builders unnoticed :open_mouth:

tspmelo replied:
The HTML linter we use is not that complete, so lots of stuff is still only caught manually

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/constants/app.constants.ts:
```diff
@@ -1,5 +1,95 @@
+import { Injectable } from '@angular/core';
+import { I18n } from '@ngx-translate/i18n-polyfill';
```
```suggestion

import { I18n } from '@ngx-translate/i18n-polyfill';
```

epuertat replied:
Done.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/url-builder.service.ts:
```diff
@@ -0,0 +1,39 @@
+import { Location } from '@angular/common';
+import { URLVerbs } from '../constants/app.constants';
```
```suggestion

import { URLVerbs } from '../constants/app.constants';
```

epuertat replied:
Done.

epuertat commented:
jenkins retest make check

epuertat commented:
jenkins test make check

epuertat commented:
jenkins test dashboard

epuertat commented:
> Working fine for me. Maybe we can add some instructions to Hacking.rst to add these rules in new code.

Very good point, @a2batic! I'll do that.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -307,17 +304,18 @@ <h3 class="page-header" i18n>Striping</h3>
+            [form]="formDir"
+            (submitAction)="submit()"
+            i18n="form action button|Example: Create Pool@@formActionButton"
+            type="button">{{ action | titlecase }} {{ resource | titlecase }}</cd-submit-button>
```
Haha, don't worry. It makes a lot of sense! I was wrongly assuming that `titlecase` pipe wouldn't covert all-caps words to lower-case.

tspmelo replied:
Is the same situation not applied to the `action`?

epuertat replied:
I'd say it's not, as long as `action` is a word `Create`, `Delete`, etc.

epuertat commented:
jenkins test docs

epuertat commented:
> I found the following error when trying to move an RBD image to Trash:
> 
> ![Screenshot from 2019-04-01 10-32-00](https://user-images.githubusercontent.com/14297426/55317965-89b2b380-5469-11e9-8542-98c8ed0defd5.png)

Thanks for pointing this out, @ricardoasmarques. I'll check this issue!

epuertat commented:
@ricardoasmarques fixed. I removed all `EntryComponents` in block and rgw modules, as I saw (some of them) were unnecessary, but it seems that is needed for components not instantiated in either templates or routes, so I reverted those 2 changes.

epuertat commented:
please jenkins test this

epuertat commented:
Jenkins retest this please

ricardoasmarques commented:
1. Apparently you missed the iSCSI and NFS actions, since those are still names 'Add' instead of 'Create':

![Screenshot from 2019-04-01 13-08-15](https://user-images.githubusercontent.com/14297426/55329552-c93bc880-5486-11e9-8159-61e7bdacf013.png)

![Screenshot from 2019-04-01 13-07-23](https://user-images.githubusercontent.com/14297426/55329566-cfca4000-5486-11e9-97cb-fe3973c3c053.png)

2. Also, something that I'm not sure if should be part or this PR, but sometimes the "object" that we are acting on starts with uppercase and sometimes not. Shouldn't be uniformed? e.g.:

![Screenshot from 2019-04-01 12-59-04](https://user-images.githubusercontent.com/14297426/55329627-fb4d2a80-5486-11e9-9ac4-64e1973ab597.png)
 VS
![Screenshot from 2019-04-01 13-03-02](https://user-images.githubusercontent.com/14297426/55329632-fee0b180-5486-11e9-8c6c-93cb99e3918b.png)

epuertat commented:
jenkins please retest this

epuertat commented:
retest this please

epuertat commented:
> Apparently you missed the iSCSI and NFS actions, since those are still named 'Add' instead of 

Yes, it was intentional and it's stated in the PR desc. When I started working on this, both iSCSI and NFS were actively modified and didn't want to collide with those. I assumed refactoring the main 8 components would be enough as a starter. As this becomes bigger and bigger, the chances to conflict with new PRs increase (e.g.: #27164 back-button component by @Devp00l, @tspmelo Bootstrap update, ... ), so I preferred to put a stop line there.

On the other hand, as I'm not familiar with iSCSI terms, we should ensure that, for example `targets` are actually created (new components) or added (existing resources included in some collection).

> Also, something that I'm not sure if should be part or this PR, but sometimes the "object" that we are acting on starts with uppercase and sometimes not. Shouldn't be uniformed? e.g.:

I did it initially by applying `titlecase` pipe to resource names as well (`{{ resource | titlecase }}`), but @tspmelo spotted that all-caps resources like `RBD` would end up as `Create Rbd`, which looks clearly wrong, and I reverted that change. So I'm going to create a new titlecase pipe that does not switch to lowercase the remaining chars (Angular source code for `titlecase` pipe):

```ts
@Pipe({name: 'titlecase'})
export class TitleCasePipe implements PipeTransform {
  /**
   * @param value The string to transform to title case.
   */
  transform(value: string): string {
    if (!value) return value;
    if (typeof value !== 'string') {
      throw invalidPipeArgumentError(TitleCasePipe, value);
    }


    return value.replace(
        unicodeWordMatch, (txt => txt[0].toUpperCase() + txt.substr(1).toLowerCase()));
  }
}
```

Is that ok @ricardoasmarques ? Thanks!

epuertat commented:
jenkins test make check

ricardoasmarques commented:
> Is that ok @ricardoasmarques ? Thanks!

Sound good

epuertat commented:
> > Is that ok @ricardoasmarques ? Thanks!
> 
> Sound good

No need of creating a new pipe. It seems that @votdev already exposed `lodash.upperFirst()` in an `upperfirst` pipe (87ab0dac19f94784b0a7181f7bbb81432e218f18). Thanks, @votdev! :smile:

epuertat commented:
jenkins test make check

epuertat commented:
jenkins test dashboard

epuertat commented:
jenkins test dashboard

epuertat commented:
jenkins test dashboard

Devp00l commented:
I'm not sure but I guess the modals in the RGW user form doesn't follow the new wording style. It should also be "Remove" on the tooltip that currently says "Delete".

![2019-04-02_10-54-11](https://user-images.githubusercontent.com/16167865/55389693-0f039a00-5536-11e9-8b18-5c6f18ead40a.png)

![2019-04-02_10-58-10](https://user-images.githubusercontent.com/16167865/55389863-58ec8000-5536-11e9-8287-329905f7576b.png)

epuertat commented:
@Devp00l if you wan't to skip the whole review, [these are](https://github.com/ceph/ceph/compare/1d4189b0a7628c4db2e6612e3dd0e865c2ffb185..debba4b99f39db4d3bb7e6049278ad2c25efe88a) just the changes you requested. Thanks for the review!

Devp00l commented:
Sorry you forgot "Capabilities".
![2019-04-02_15-56-15](https://user-images.githubusercontent.com/16167865/55408948-52bfc900-5561-11e9-8ba5-d4f2082464d1.png)

epuertat commented:
> Sorry @epuertat, you seem to have forgotten "Capabilities".
> ![](https://user-images.githubusercontent.com/16167865/55408948-52bfc900-5561-11e9-8ba5-d4f2082464d1.png)

@Devp00l Hehe, that was an intentional omission (but anyway subject to further discussion): my assumption here is that 'Capabilities' fall into the category of items that can be 'added' to another object rather than 'created' from scratch, as they basically consist of existing elements (`read`/`write` permissions, and `types`).

For example, when you 'create' an S3 Key, you're actually defining and storing a new instance of that, but when you 'create' a capability you are rather re-arranging things that already existed.

epuertat commented:
jenkins test make check arm64

epuertat commented:
:warning:  Fix https://github.com/ceph/ceph/pull/27406 is needed to backport with this one.
